opm/output/eclipse/Summary.hpp

```diff
@@ -21,66 +21,58 @@
-#include <ert/util/ert_unique_ptr.hpp>
-
-#include <opm/parser/eclipse/EclipseState/Grid/EclipseGrid.hpp>
-#include <opm/parser/eclipse/EclipseState/Schedule/SummaryState.hpp>
-
-#include <opm/output/data/Wells.hpp>
-#include <opm/output/eclipse/RegionCache.hpp>
-
 namespace Opm {
-
+    class EclipseGrid;
     class EclipseState;
     class Schedule;
     class SummaryConfig;
-    class EclipseGrid;
-    class Schedule;
+    class SummaryState;
+} // namespace Opm
 
-namespace out {
+namespace Opm { namespace data {
+    class WellRates;
+}} // namespace Opm::data
+
+namespace Opm { namespace out {
 
 class Summary {
 public:
-    Summary( const EclipseState&, const SummaryConfig&, const EclipseGrid&, const Schedule& );
-    Summary( const EclipseState&, const SummaryConfig&, const EclipseGrid&, const Schedule&, const std::string& );
-    Summary( const EclipseState&, const SummaryConfig&, const EclipseGrid&, const Schedule&, const char* basename );
-
-    void add_timestep(const SummaryState& st,
-                      int report_step);
+    using GlobalProcessParameters = std::map<std::string, double>;
+    using RegionParameters = std::map<std::string, std::vector<double>>;
+    using BlockValues = std::map<std::pair<std::string, int>, double>;
 
+    Summary(const EclipseState&  es,
+            const SummaryConfig& sumcfg,
+            const EclipseGrid&   grid,
+            const Schedule&      sched,
+            const std::string&   basenm = "");
 
-    void eval(SummaryState& summary_state,
-              int report_step,
-              double secs_elapsed,
-              const EclipseState& es,
-              const Schedule& schedule,
-              const data::Wells&,
-              const std::map<std::string, double>& single_values,
-              const std::map<std::string, std::vector<double>>& region_values = {},
-              const std::map<std::pair<std::string, int>, double>& block_values = {}) const;
+    ~Summary();
 
+    void add_timestep(const SummaryState& st, const int report_step);
 
+    void eval(SummaryState&                  summary_state,
+              const int                      report_step,
+              const double                   secs_elapsed,
+              const EclipseState&            es,
+              const Schedule&                schedule,
+              const data::WellRates&         well_solution,
+              const GlobalProcessParameters& single_values,
+              const RegionParameters&        region_values = {},
+              const BlockValues&             block_values  = {}) const;
 
-    ~Summary();
     void write() const;
 
 private:
-    void internal_store(const SummaryState& summary_state, int report_step);
-
-
-    class keyword_handlers;
-
-    const EclipseGrid& grid;
-    out::RegionCache regionCache;
-    ERT::ert_unique_ptr< ecl_sum_type, ecl_sum_free > ecl_sum;
-    std::unique_ptr< keyword_handlers > handlers;
+    class Impl;
+    std::unique_ptr<Impl> pImpl_;
```

Comment:
No please - not these opaque `Impl` classes and symbols.

Reply:
I'll just comment on this briefly: I agree for the most part, but if it lets you avoid including tons of headers in this class' header it can be a very good idea, improving compile times a lot (not sure that is the case here though).

Reply:
In this case we're operating under somewhat constrained circumstances.  My primary concern, which indeed is a subjective Call, is that I don't want knowledge about the big list of function pointers to leak out of `Summary.cpp`.

If we're not going to use pointer-to-implementation, then we need to expose the `Impl` class's [list of data members and types](https://github.com/OPM/opm-common/blob/a9ce5a67dafaf373b10fcf1d043868e9a4a47dce/src/opm/output/eclipse/Summary.cpp#L1990-L2019) to all who use (i.e., `#include`) `Summary.hpp`.  Most of the types are already public so it's not that much extra information.

On the other hand, the [`SummaryOutputParameters`](https://github.com/OPM/opm-common/blob/a9ce5a67dafaf373b10fcf1d043868e9a4a47dce/src/opm/output/eclipse/Summary.cpp#L1815-L1859) and the [`SMSPecStreamDeferredCreation`](https://github.com/OPM/opm-common/blob/a9ce5a67dafaf373b10fcf1d043868e9a4a47dce/src/opm/output/eclipse/Summary.cpp#L1861-L1890) types don't really make a lot of sense outside of the private implementation details of class `Summary`, so I'd really prefer if they were not exposed.

In the end it's a judgement call.  My preference here is to keep the pointer-to-implementation mechanism, at least for now.

Reply:
My main gripe is with the class/instance name `Impl` which is world leading non-informative name

Reply:
Would it be better if I renamed it to
```
class InternalImplementationDetail
```
or something of the sort?

Reply:
SummaryImplementation?

Reply:
>> Would it be better if I renamed it to
>> ```
>> class InternalImplementationDetail
>> ```
>> 
>> or something of the sort?
>
> ```
> SummaryImplementation?
> ```

Good idea.  I'll update accordingly.  Thanks for the suggestion.